### PR TITLE
lp:1438990 Abort in slave if SAVEPOINT is in trigger.

### PR DESCRIPTION
**Pull request by vlad-lesin**
1. Why does the crash take place?

The crash happens in Write_rows_log_event::write_row()  when row field is unpacked. The stack in the following

Relay_log_info::get_table_data()
unpack_row()
unpack_current_row()
Write_rows_log_event::write_row()

Relay_log_info::get_table_data() contains the following code:

  bool get_table_data(TABLE *table_arg, table_def tabledef_var, TABLE conv_table_var) const
  {
     ...
    for (TABLE_LIST *ptr= tables_to_lock ; ptr != NULL ; ptr= ptr->next_global)
      if (ptr->table == table_arg)
      {
         ...
         return true;
      }
     return false;
   }

Relay_log_info::tables_to_lock is NULL, Relay_log_info::get_table_data() returns false and unpack_row() is aborted by assert.
1. Why Relay_log_info::tables_to_lock is NULL.?

Because it was cleared in Query_log_event::do_apply_event()->Relay_log_info::slave_close_thread_tables()->Relay_log_info::clear_tables_to_lock() when log event for "SAVEPOINT" is executed"
1. How table map is written for trigger?

It is written once for the whole trigger body:

a) Table is added to thd->lex->query_tables in open_table_entry_fini() <- the function for opening tables for triggers which are specified for the table which takes part in query

b) thd->lock is filled from lock_tables() with thd->lex->query_tables . lock_tables .

c) binlog_log_row()->write_locked_table_maps() writes table maps to binlog and invoke THD::binlog_write_table_map() which increases THD::binlog_table_maps.

d) when write_locked_table_maps() is invoked for the next statement in trigger it writes map to binlog only if THD::binlog_table_maps is zero, but it is not zero as it was changed in (c).

Trigger is parsed and executed before DML execution and all tables which take part in query are added into list of open tables.
1. The fix:

SAVEPOINT, ROLLBACK TO SAVEPOINT(as well as BEGIN, COMMIT, ROLLBACK, but they are forbidden for triggers) statements are logged into binary log as Query_log_event. Query_log_event::do_apply_event() wipes out table map.  And we can not recognize on slave if we are inside of trigger or not as only modification events are written to binary log and there is not event for trigger boddy call. But we can recognize it in master. The idea is to write tables map into binary log after each SAVEPOINT, ROLLBACK TO SAVEPOINT execution on master if it happens inside of trigger.

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/838/

### CODE REVIEW

**Comment by laurynas-biveinis** [comment]
The basic fix idea (write out extra table map events on the master) looks good. But what about the case where ROLLBACK TO SAVEPOINT / SAVEPOINT is the last statement on the trigger? Won't we write this new extra table map event, and then write the table map event for the next SQL statement, which shows that the first event is redundant? If so, can we avoid this?

For the testcase, I'd add
1) an explicit slave SQL sync with master between last insert and first drop table, for the slave abort before the fix;
2) show binlog events on the master to see the new table map event;
3) add the case above (SAVEPOINT/ROLLBACK TO SAVEPOINT last in the trigger) and show its events too.

**Comment by vlad-lesin** [comment]
> Won't we write this new extra table map event, and then write the table map event for the next SQL statement, which shows that the first event is redundant? 

I checked this case. Yes, we will have redundant map info at the end of trigger body but it will be wiped out either at the end of the statement for which the trigger was invoked or at the beginning of the next statement on slave. 

> If so, can we avoid this?

I think no. At least I have not found the way to understand if the current statement is the last in trigger body.

I agree with necessity to test those edge cases.

**Comment by laurynas-biveinis** [comment]
OK. I am slightly worried whether the slave processes two table map events in a row correctly, if that doesn't result in i.e. a memory leak. Do you know? 

I see that https://bugs.mysql.com/bug.php?id=37137 mentioning that two identical table map events in a row are benign.

**Comment by vlad-lesin** [comment]
> OK. I am slightly worried whether the slave processes two table map events in a row correctly, if that doesn't result in i.e. a memory leak. Do you know? 

There are no two table maps. Table map is written once per whole trigger body. SAVEPOINT clean this table in slave and then slave loads this table again from relay log as master wrote it just after SAVEPOINT event.

On slave table map is cleared by Rows_log_event::do_apply_event() just after execution of the statement for which trigger was inoked. Here is the backtrace:

(gdb) bt
#0  Relay_log_info::clear_tables_to_lock (this=0x1b95b40)
    at ./sql/rpl_rli.cc:1557
#1  0x0000000000a77cf3 in Relay_log_info::slave_close_thread_tables (
    this=0x1b95b40, thd=0x7fffc8000a40)
    at ./sql/rpl_rli.cc:1631
#2  0x0000000000a779b8 in Relay_log_info::cleanup_context (this=0x1b95b40, 
    thd=0x7fffc8000a40, error=false)
    at ./sql/rpl_rli.cc:1534
#3  0x0000000000a1c38f in rows_event_stmt_cleanup (rli=0x1b95b40, 
    thd=0x7fffc8000a40)
    at ./sql/log_event.cc:11684
#4  0x0000000000a1c182 in Rows_log_event::do_apply_event (this=0x7fffc8041400, 
    rli=0x1b95b40)
    at ./sql/log_event.cc:11581
#5  0x0000000000a06a91 in Log_event::apply_event (this=0x7fffc8041400, 
    rli=0x1b95b40)
    at ./sql/log_event.cc:3125
#6  0x0000000000a615cc in apply_event_and_update_pos (ptr_ev=0x7ffff02e08b0, 
    thd=0x7fffc8000a40, rli=0x1b95b40)
    at ./sql/rpl_slave.cc:3783
#7  0x0000000000a62b06 in exec_relay_log_event (thd=0x7fffc8000a40, 
    rli=0x1b95b40)
    at ./sql/rpl_slave.cc:4280
---Type <return> to continue, or q <return> to quit---
#8  0x0000000000a68374 in handle_slave_sql (arg=0x1b47d70)
    at ./sql/rpl_slave.cc:6163
#9  0x0000000000d9e879 in pfs_spawn_thread (arg=0x7fffcc014c00)
    at ./storage/perfschema/pfs.cc:1860
#10 0x00007ffff737d182 in start_thread (arg=0x7ffff02e1700)
    at pthread_create.c:312
#11 0x00007ffff688a47d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

**Comment by vlad-lesin** [comment]
I would like to comment the output of bilog events.

When trigger is invoked the all tables which can be changed in this trigger is written in binlog table map. And there is no table map per statement inside of trigger. EXCEPT procedure call. When such call takes place new table map for tables which can be changed in procedure is written to the binary log. But this can not the cause of memory leak on slave as table map is just a list of pairs and adding table to map is just single event in binary log (Table_map_log_event::do_apply_event()), and this list is cleared after statement execution.

The test is the following:

CREATE TRIGGER tr1 AFTER INSERT ON t1 FOR EACH ROW
BEGIN  
        SAVEPOINT event_logging;
        INSERT INTO t2 VALUES (1);
        RELEASE SAVEPOINT event_logging;
        INSERT INTO t3 VALUES (2);
        # Test if the fix works for nested SP call
        CALL p1();
END|

p1() inserts rown in t2 and t3. And here is the binlog:

+master-bin.000001      #       Query   #       #       BEGIN
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t1)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t2)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t3)

This map is for trigger.

+master-bin.000001      #       Table_map       #       #       table_id: # (test.t2)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t3)

This map is for call p1().

+master-bin.000001      #       Write_rows      #       #       table_id: #

Insert.

+master-bin.000001      #       Query   #       #       SAVEPOINT `event_logging`

Savepoint writes all tables from table map, i.e for trigger and for procedure call.

+master-bin.000001      #       Table_map       #       #       table_id: # (test.t1)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t2)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t3)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t2)
+master-bin.000001      #       Table_map       #       #       table_id: # (test.t3)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ SAVEPOINT

+master-bin.000001      #       Write_rows      #       #       table_id: # flags: STMT_END_F
+master-bin.000001      #       Xid     #       #       COMMIT /\* XID */

**Comment by vlad-lesin** [comment]
> But what about the case where ROLLBACK TO SAVEPOINT / SAVEPOINT is the last statement on the trigger? 

And by the way. I tried to set up savepoints at the end of of trigger. They will not be written to row-based binary log at all. As they do not cause any row changes. The same is true for implicit ROLLBACK TO. Only changes over rows are written to binary log. 

But if "DECLARE EXIT HANDLER FOR SQLEXCEPTION" is defined and there was exception then SAVEPOINT is logged to binary log.

**Comment by vlad-lesin** [comment]
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/845/

File: sql/handler.cc [hunk 1]
```diff
@@ -2136,6 +2138,14 @@ int ha_rollback_to_savepoint(THD *thd, SAVEPOINT *sv)
     status_var_increment(thd->status_var.ha_savepoint_rollback_count);
     trans->no_2pc|= ht->prepare == 0;
   }
+
+  /*
+    Write tables map once more for trigger as it will be wiped out in
+    Write_rows_log_event::write_row() on slave.
+  */
+  if (thd->in_sub_stmt == SUB_STMT_TRIGGER)
+    error= write_locked_table_maps(thd, true);
```
Comment (laurynas-biveinis) [inline sql/handler.cc:2147]: One minor issue re. error flag, but I'm not sure whether it's worth fixing. ha_rollback_to_savepoint sets it disregarding any previous value in the flag for each ht->*_rollback call. The failure of these calls might as well be an assert failure only, as the comments "cannot happen" say, because the main way they would fail in InnoDB is if the savepoint name is not found, which is validated earlier. (Not sure about TokuDB).

But with this new call the rollback may also fail if we have non-trx tables in the binlog event group and we get a disk full error while writing the ROLLBACK event to binlog, which is quite unlikely but not a "cannot happen" error.

If you think it's worth addressing, s/error= /error|= /g, otherwise merge it or let me know I should merge it.

File: sql/handler.cc [hunk 2]
```diff
@@ -2231,6 +2241,14 @@ int ha_savepoint(THD *thd, SAVEPOINT *sv)
     }
     status_var_increment(thd->status_var.ha_savepoint_count);
   }
+
+  /*
+     Write tables map once more for trigger as it will be wiped out in
+     Write_rows_log_event::write_row() on slave.
+   */
+   if (thd->in_sub_stmt == SUB_STMT_TRIGGER)
+     error= write_locked_table_maps(thd, true);
```
Comment (laurynas-biveinis) [inline sql/handler.cc:2250]: Likewise

**Comment by laurynas-biveinis** [comment]
Merging as-is, the issue is too minor to hold the release

File: sql/handler.cc (hunk 1 above)
Comment (vlad-lesin) [inline sql/handler.cc:2147]: > One minor issue re. error flag, but I'm not sure whether it's worth fixing. ha_rollback_to_savepoint sets it disregarding any previous value in the flag for each ht->*_rollback call. 

Yes, the logic is if only one storage engine returns error then the whole rollback procedure returns error. As error de-facto is boolean here then "error=" and "error|=" do the same thing.

> The failure of these calls might as well be an assert failure only, as the comments "cannot happen" say, because the main way they would fail in InnoDB is if the savepoint name is not found, which is validated earlier

I would not believe this comment. This is common code which must work for any engine and some abstract engine in vacuum can have any reason to fail rollback procedure, and this code must take it into account.

About innodb. What about this?
trx_rollback_to_savepoint_for_mysql() (): 
    switch (trx->state) {
    case TRX_STATE_NOT_STARTED:
        ut_print_timestamp(stderr);
        fputs("  InnoDB: Error: transaction has a savepoint ", stderr);
        ut_print_name(stderr, trx, FALSE, savep->name);
        fputs(" though it is not started\n", stderr);
        return(DB_ERROR);

I understand your doubts. I ran down to the call stack:
ha_rollback_to_savepoint()
trans_rollback_to_savepoint()
mysql_execute_command()

And error in ha_rollback_to_savepoint() is processed correctly.

Comment (laurynas-biveinis) [inline sql/handler.cc:2147]: Thanks Vlad, you are right
